aws/src/main/java/org/apache/iceberg/aws/s3/S3LocationProvider.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.aws.s3;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Map;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.TableProperties;
+import org.apache.iceberg.io.LocationProvider;
+import org.apache.iceberg.relocated.com.google.common.hash.HashCode;
+import org.apache.iceberg.relocated.com.google.common.hash.HashFunction;
+import org.apache.iceberg.relocated.com.google.common.hash.Hashing;
+import org.apache.iceberg.util.LocationUtil;
+
+/**
+ * This location provider provides data locations that are optimized for S3 performance. Both
+ * General Purpose buckets and Directory buckets will see better throughput and autoscaling behavior
+ * than using the generic ObjectStoreLocationProvider.
+ *
+ * <p>The data location is resolved as follows. Data files are written directly at this path with no
+ * other intermediate directories created.
+ *
+ * <ol>
+ *   <li>{@link TableProperties#WRITE_DATA_LOCATION}
+ *   <li><code>tableLocation + "/data"</code>
+ * </ol>
+ *
+ * The data file is placed immediately under the data location. Partition names are <b>not</b>
+ * included. The data filename is prefixed with a 24-character binary hash, which ensures that files
```

Comment:
The `24-character` prefix and lack of a directly marker (`/`) is a little problematic for older clients (like HadoopFS) and procedures like orphan files because anything that needs to list that directory will run into problems with the key space.  We already have an issue with this with the current provider, but this would compound that problem since there's no hierarch. 

This would make it a little easier to brake up prefix-separated ranges for listing (especially due to the low character limit), but that's something would need to look into on the orphan files action.

Reply:
Is the problem coming from the delimited list or is it some type parallelization issue due to lack of directories? I was thinking not having the directory essentially remove the requirement for discovering them and it can be just a flat list to discover everything under the write.data.path.

Reply:
The issue is having too much cardinality for FS implementations that think in terms of directory structure.  By breaking this up we can address both scenarios (files system based and object store listing) at the same time.